File: _site/wireworld_compiler.rb
myCircuit = [:gate, [:gate, [:var], [:var]], [:gate, [:var], [:var]]]

class BoardItem
  attr_accessor :x, :y
end

class Gate < BoardItem
  def initialize(direction, x, y)
    @direction = direction
    @x = x
    @y = y
  end

  def inspect
    "<gate #{@direction} (#{@x},#{@y})>"
  end
end

class Var < BoardItem
  def initialize(direction, x, y)
    @direction = direction
    @x = x
    @y = y
  end

  def inspect
    "<var #{@direction} (#{@x},#{@y})>"
  end
end

class Wire < BoardItem
  def initialize(dir1, dir2, x, y)
    @dir1 = dir1
    @dir2 = dir2
    @x = x
    @y = y
  end

  def inspect
    "<wire #{@dir1} #{@dir2} (#{@x},#{@y})>"
  end
end

def height(circuit)
  circuit.map(&:y).max - circuit.map(&:y).min + 1
end

def move(circuit, dx, dy)
  circuit.each do |component|
    component.x += dx
    component.y += dy
  end
end

def compileCircuit(circuit, direction = :down)
  case circuit.first
  when :var
    [Var.new(direction, 0, 0)]
  when :gate
    l = compileCircuit(circuit[1], :down)
    r = compileCircuit(circuit[2], :up)
    size_l = height(l)
    size_r = height(r)
    move(r, 0, size_l)

    wiringAcross = (size_l...size_l + size_r - 1).map { |x| Wire.new(:up, :down, x, 0) }
    wiringUp = (size_r...size_l + size_r - 1).map { |x| Wire.new(:down, :up, x, size_l + size_r - 1 - x) }
    finalGate = Gate.new(direction, size_l + size_r - 1, 0)
    return l + r + wiringAcross + wiringUp + [finalGate]
  end
end

size2 = [Var.new(:down, 0, 0), Var.new(:down, 0, 1), Gate.new(:down, 1, 0)]
size1 = [Var.new(:down, 0, 0)]

p compileCircuit([:gate, [:gate, [:var], [:var]], [:gate, [:var], [:var]]])

File: _site/wireworld_compiler.js
$(function () {
  var example = ["binop", "and", ["var", "x"], ["var", "y"]];

  var height= function(stuff) {
    var max = 0;
    for (var i = stuff.length - 1; i >= 0; i--) {
      max = max > stuff[i].x ? max : stuff[i].x
    };
    return max + 1;
  }

  var moveDown = function (circuit, dy) {
    for (var i = circuit.length - 1; i >= 0; i--) {
      circuit[i].y += dy;
    };
  }

  var compileExpression = function (expr, dir){
    if (!dir) {
      dir = "down";
    }

    if (expr[0] == "var") {
      return [{component_type: "var", dir: dir, x:0, y:0}];
    } else if (expr[0] == "binop") {
      var lhs = compileExpression(expr[2], "down");
      var rhs = compileExpression(expr[3], "up");
      var size_l = height(lhs);
      var size_r = height(rhs);
      moveDown(rhs, size_l);

      out = [];
      out.concat(lhs);
      out.concat(rhs);

      for (var i = size_l; i < size_l + size_r - 1; i++) {
        out.push({component_type: "wire",
                  start_dir: "up",
                  end_dir: "down",
                  x: i,
                  y: 0});
      };

      for (var i = size_r; i < size_l + size_r - 1; i++) {
        out.push({component_type: "wire",
                  start_dir: "down",
                  end_dir: "up",
                  x: i,
                  y: size_l + size_r - 1 - i});
      };

      out.push({component_type: "gate",
                  gate_type: expr[1],
                  dir: dir,
                  x: size_l + size_r - 1,
                  y: 0})

      return out;

    }
  }
});